File: Defender/ArgumentTypeException.cs
using System;

namespace Defender {
    /// <summary>
    /// The exception that is thrown when an object is passed to a method that expects a specific type, but is not of that type.
    /// </summary>
    public class ArgumentTypeException : ArgumentException {
        public ArgumentTypeException() { }
        public ArgumentTypeException(String message) : base(message) { }
        public ArgumentTypeException(String message, Exception inner) : base(message, inner) { }
        public ArgumentTypeException(String paramName, String message) : base(message, paramName) { }
        public ArgumentTypeException(String paramName, String message, Exception innerException) : base(message, paramName, innerException) { }
        public ArgumentTypeException(String paramName, Type paramType) : this(paramName, $"Argument not of type {paramType}.") { }
        public ArgumentTypeException(String paramName, Type paramType, String message) : this(paramName, $"Argument not of type {paramType}. {message}") { }
        public ArgumentTypeException(String paramName, Type paramType, Exception innerException) : this(paramName, $"Argument not of type {paramType}.", innerException) { }
        public ArgumentTypeException(String paramName, Type paramType, String message, Exception innerException) : this(paramName, $"Argument not of type {paramType}. {message}", innerException) { }
    }
}

File: Defender/ArgumentTypeException.vb
''' <summary>
''' The exception that Is thrown when an object Is passed to a method that expects a specific type, but Is Not of that type.
''' </summary>
Public Class ArgumentTypeException
    Inherits ArgumentException
    Public Sub New()
        MyBase.New()
    End Sub

    Public Sub New(Message As String)
        MyBase.New(Message)
    End Sub

    Public Sub New(Message As String, Inner As Exception)
        MyBase.New(Message, Inner)
    End Sub

    Public Sub New(ParamName As String, Message As String)
        MyBase.New(Message, ParamName)
    End Sub

    Public Sub New(ParamName As String, Message As String, Inner As Exception)
        MyBase.New(Message, ParamName, Inner)
    End Sub

    Public Sub New(ParamName As String, ParamType As Type)
        MyBase.New(ParamName, $"Argument not of type {ParamType}.")
    End Sub

    Public Sub New(ParamName As String, ParamType As Type, Message As String)
        Me.New(ParamName, $"Argument not of type {ParamType}. {Message}")
    End Sub

    Public Sub New(ParamName As String, ParamType As Type, Inner As Exception)
        Me.New(ParamName, $"Argument not of type {ParamType}.", Inner)
    End Sub

    Public Sub New(ParamName As String, ParamType As Type, Message As String, Inner As Exception)
        Me.New(ParamName, $"Argument not of type {ParamType}. {Message}", Inner)
    End Sub
End Class
